src/OpenApi/src/Transformers/DelegateOpenApiDocumentTransformer.cs

```diff
@@ -0,0 +1,46 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.OpenApi.Any;
+using Microsoft.OpenApi.Models;
+
+namespace Microsoft.AspNetCore.OpenApi;
+
+internal sealed class DelegateOpenApiDocumentTransformer : IOpenApiDocumentTransformer
+{
+    private readonly Func<OpenApiDocument, OpenApiDocumentTransformerContext, CancellationToken, Task>? _documentTransformer;
+    private readonly Func<OpenApiOperation, OpenApiOperationTransformerContext, CancellationToken, Task>? _operationTransformer;
+
+    public DelegateOpenApiDocumentTransformer(Func<OpenApiDocument, OpenApiDocumentTransformerContext, CancellationToken, Task> transformer)
+    {
+        _documentTransformer = transformer;
+    }
+
+    public DelegateOpenApiDocumentTransformer(Func<OpenApiOperation, OpenApiOperationTransformerContext, CancellationToken, Task> transformer)
+    {
+        _operationTransformer = transformer;
+    }
+
+    public async Task TransformAsync(OpenApiDocument document, OpenApiDocumentTransformerContext context, CancellationToken cancellationToken)
+    {
+        if (_documentTransformer != null)
+        {
+            await _documentTransformer(document, context, cancellationToken);
+        }
+
+        if (_operationTransformer != null)
+        {
+            var documentService = context.ApplicationServices.GetRequiredKeyedService<OpenApiDocumentService>(context.DocumentName);
+            foreach (var pathItem in document.Paths.Values)
+            {
+                foreach (var operation in pathItem.Operations.Values)
+                {
+                    var descriptionId = ((OpenApiString)operation.Extensions[OpenApiConstants.DescriptionId]).Value;
```

Comment:
Is this guaranteed to be there and be this type?

Reply:
Good question!

There is a risk that the description ID might not be there if the a previous transformer has run and removed the extension attribute. I'm updating the code to throw an exception in this case. We could alternatively chose to do a slow-path construction of the context object but I'd much rather be stringent about not doing anything to change the caching behavior.